thegroundwork/groundwork.js
# Project Cleanup

Went through and polished / cleaned the repo. Notably:
- Moved core commands to `Makefile` and consolidated / organized. See https://github.com/thegroundwork/groundwork.js/blob/cp-cleanup/Makefile
- Added more logging
- Better automation. When running `npm run release`, you bump version in CLI which will send a new branch to github. Travis then builds docs and examples and afterwards syncs to S3. We no longer have to manually build individual tasks before releasing, which cuts down on a lot of steps. And no longer have to commit `docs` or `dist` to repo either, instead triggering their generation via a `postinstall` `package.json` hook for the user. 
- Simplified development / examples story -- everything is now contained under `npm start`.  This builds the library and launches a server at `http://localhost:3000` and `http://localhost:3000/examples` 
- Slight restructure in CDN files / organization, which allows us to simplify example `<script>` src's down to only relative filepaths, thereby removing need for `transformer.js`. The root (currently a test bucket in gw-dev -- `cdn.dev.thegroundwork.com/thegroundwork`) now contains the `latest` version of Groundwork.js, docs and examples for easy reference across services without the need for specific version numbers, and then within that folder there are individual versioned folders `cdn.dev.thegroundwork.com/thegroundwork/1.1.1` that contain docs and examples specifically for that version. This allows us to more completely track changes over time by grouping docs and examples together with their corresponding sources. So
  - `http://cdn.thegroundwork.com/groundworkjs` - path to `latest`
  - `http://cdn.thegroundwork.com/groundworkjs/1.1.1` - also path to `latest` 
  - `http://cdn.thegroundwork.com/groundworkjs/1.1.0` - path to 1.1.0
  - `http://cdn.thegroundwork.com/groundworkjs/1.1.0/examples` - examples for 1.1.0
  - `http://cdn.thegroundwork.com/groundworkjs/0.9.0/doc` - docs for 0.9.9
 - ...and so on.
- Add Babel transforms around Lodash for plucking only what is needed; filesize is a bit lower.
- Parse examples with `html-webpack-plugin` so that we can dynamically update filenames, as well as use `html-loader` for sharing partials, thereby removing duplicate code across files 
- Lots of subtle cleanup all around

To test
-------
- Since I removed `dist` and `docs` from repo, too many changes to track in files pane. Go here https://github.com/thegroundwork/groundwork.js/tree/cp-cleanup or check out locally and review
- Login to `gw-dev` and find the S3 bucket `cdn.dev.thegroundwork.com`
- Run `npm run release` and bump version number, making sure there isn't already a tag in place on github
- Once complete, inspect Travis log output
- See latest here: http://cdn.dev.thegroundwork.com/groundworkjs/
- Return to S3 and note the new folder, as well as updated root by clicking into groundwork.js and inspecting banner
- Delete test tag and branch from github and local, revert package.json to `1.1.1`

Todo
-----
- [x] Figure out how to get S3 bucket tied to real, accessible `cdn.dev.thegroundwork.com` public route. Something seems to have changed somewhere in AWS (or not?). @tylergaw - can I get your help with this? (**NOTE**: Tried to use Frontier, but couldn't figure out how to add a new project. Could also use some help around this)
- [x] Ensure examples work with relative paths across various version numbers 
- [x] Update main html page to show which GW.js version is latest
- [x] Update "groundwork.js is available..." console message to include version reference
- [x] Investigate cloudflare caching since we're now pushing latest into root. We need a way to bust the cache. Will follow up with @ewdurbin when i get back from wedding on Tuesday. 
- [x] Update release.sh and remove `test` from dynamic branch name 
- [x] Update bucket to `gw-prod` in `Makefile`.
- [x] Update `.travis.yml` keys to prod: `AWS_ACCESS_KEY_ID=AKIAJ2FJKT7KSNFRPWCQ` and `"AFB8OtuKeqmbkCAwHmthc78yQd+o62twi4sTiNAi7OhRiQMYYxYQeNee3tJSVXOvnQ/nGyat7X5oQFVx0mWm560eQwE/8XyZPKI1tfT9Wd0zOYYprV7m8rMg1sxaEI3zwbF+f2JDcZxY6xvV2isXmGlYF6fJNyIig3GPBTEON571Al2OXqhwXH+Ey068KAYReUf2q4WLXGea+bcFWo2+Al8PM9AqmYmqfKoIuI8muVvpcSAfJd0+nZZZr9YQvdQ+iXy1wkvlv5S3T2TTRSovAO20wyMf5ATruL0aCzPd2JZor+UKKP8sciI48cGJACqY9HBJAjxGiXgSKWeZ5oTxDXIyz/okH9RY/Bjf5Z0G4qjhFvPUdqoWcaFsnyt/E+YCTS0qc5TwgFoicR6Ug+iF94QGW2PGmd4mMV8q2G9tGJ6NiArGkdIJbjgKT1uJVou9aVCKnuvWq8rFLaj4jOVkPNX24OUo9MSzrr4JWekBA6UzNba5cfW9CMQ+MvLmOt2LPDZYiDeVxeW2Kphi4xly8X+0hzf0Z6YvKN6qe8tJHCO0E5j9VRSWGRMAOKoZiP4X5IsFzxvIoc8Hu8slMmz2u6PZNjDTa1/v2jpE8UMSh7YewqOWf0Vv2jQsvzD+7e5MoE1adslFxGTzmoYLSLyynC7rwd9+bMuolNVJ66FQf9w="`
- [x] Update `CLOUDFRONT_DISTRIBUTION` to `E1YK0CEUDW7K3G `
